src/Microsoft.TestPlatform.Common/ExtensionFramework/Utilities/TestExtensions.cs

```diff
@@ -390,4 +479,35 @@ internal void InvalidateCache()
         }
     }
 
+    private void AddCachedExtensionToExtensionMap<T>(
+        IDictionary<string, ISet<string>> extensionMap,
+        string extensionType,
+        IList<T> extensions) where T : TestPluginInformation
+    {
+        if (extensions == null) return;
+
+        extensionMap.Add(extensionType, new HashSet<string>());
+        foreach (var extension in extensions)
+        {
+            extensionMap[extensionType].Add(extension.IdentifierData);
+        }
+    }
+
+    private static string SerializeExtensionMap(IDictionary<string, ISet<string>> extensionMap)
```

Comment:
Why is this necessary? Is this to avoid approving a lot of different data points in gdpr portal? Can a more native format be used? Like json?

Reply:
Having to approve a lot of different data points in the GDPR portal is only part of the reason. 

There's also the fact that each different extension type we add telemetry data points for would have to be represented as a separate column in the telemetry table, which causes the invalidation of the previous table schema. In short, we need to recreate and repopulate the telemetry table for each newly added column. Even if that weren't the case, we'd still have to log (possibly) multiple extensions for each extension type which would still require some parsing, so we might as well just cut off the inconvenience of recreating the table and do the parsing we would've done either way, only a little more complex.

Another reason of compacting data like this is also related to the way data collectors telemetry is handled, i.e. all telemetry is being written in the same field in the telemetry table and packed using a custom structure. Granted, we can do better here by giving JSON a try.